### PR TITLE
Add support for newest checkpoint, fix attention prior

### PR DESCRIPTION
**Pull request by vklimkov-nvidia**
Changes to fix attention prior: actually use attention mask in attention kernel. Introduce support for newest checkpoint: some values are hardcoded, we should move them to config. Set inputLength to encoderLen + 1, because attention kernel attends only inputLength - 1 timesteps in kv sequence during cross attention. Currently we achieve CER comparable to NeMo, but the end of generation detection should be still improved.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h [hunk 1]
```diff
@@ -2094,6 +2094,13 @@ __global__ void __launch_bounds__(MAX_THEADS_PER_BLOCK, MIN_BLOCKS_PER_SM) maske
                 }
             }
 
+            if (is_active && has_attention_mask && DO_CROSS_ATTENTION) {
+                // TODO: This is a fix to take into account custom attention mask during cross attention.
+                // It is implicitely excludes EOS token from encoder sequence, this is to be checked.
+                // It penalizes masked tokens with -1e9, this can be adjusted to implement attention prior floor.
+                qk_ += (1e9 * (float(attention_mask_ptr[local_time_now]) - 1.0f));
```
Comment (vklimkov-nvidia) [inline cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h:2100]: this enabled the attention mask to work for cross attention

File: cpp/tensorrt_llm/batch_manager/encoderBuffers.cpp [hunk 2]
```diff
@@ -525,8 +525,9 @@ void EncoderBuffers::fill(
             auto const encoderOutputSlice = runtime::ITensor::slice(encoderOutput, offset, size);
             manager.copy(*llmReq->getEncoderOutput(), *encoderOutputSlice);
             offset += size;
-
-            inputLengthsAll.emplace_back(size);
+            // TODO: this is to ensure that EOS is included into cross-attention computation
+            // need to check if it has negative effect
+            inputLengthsAll.emplace_back(size + 1);
```
Comment (vklimkov-nvidia) [inline cpp/tensorrt_llm/batch_manager/encoderBuffers.cpp]: this changed CER from 4.7% to 1.8%. need to confirm that it affects only the cross attention

File: tensorrt_llm/models/t5tts/model.py [hunk 3]
```diff
@@ -549,9 +554,21 @@ def forward(self,
 
         hidden_states = self.cross_attention_layernorm(hidden_states)
         encoder_output = self.cross_attention_memory_layernorm(encoder_output)
+        # TODO: unnecessary computation, can be done just once for all layers
+        # TODO: ignores cross_attention_packed_mask, but they seems to be not changed anyways
+        cross_attention_mask = attention_mask_params.cross_attention_mask
+        if not self.apply_prior:
```
Comment (vklimkov-nvidia) [inline tensorrt_llm/models/t5tts/model.py:560]: TODO: check functional.py and do more efficiently

File: cpp/include/tensorrt_llm/batch_manager/llmRequest.h [hunk 4]
```diff
@@ -1183,6 +1183,26 @@ class GenericLlmRequest
...
     }
 
+    void setFocusCrossAttentionMask(runtime::BufferManager const& manager, int size=3)
+    {
+        if (!mCrossAttentionMask.has_value()) {
+            mCrossAttentionMask = std::move(manager.emptyTensor(runtime::MemoryType::kGPU, nvinfer1::DataType::kBOOL));
+            (*mCrossAttentionMask)->reshape(runtime::ITensor::makeShape({1, getEncoderOutputLen()}));
+        }
+        // Create array of booleans
+        bool* mask = new bool[getEncoderOutputLen()];
```
Comment (tijyojwad) [inline cpp/include/tensorrt_llm/batch_manager/llmRequest.h]: better to use smart pointers like unique_ptr if it's used within the same scope

File: cpp/include/tensorrt_llm/batch_manager/llmRequest.h [hunk 5]
```diff
@@ -1769,25 +1789,27 @@ class GenericLlmRequest
...
         if (attentionPriorIdx >= getEncoderOutputLen() - 5) {
             mAttentionPriorCounterCloseToEnd++;
         }
     }
 
-    bool isAttentionPriorStuck() const
+    bool isAttentionPriorFinished() const
     {
-        return mAttentionPriorCounter >= 10;
+        bool isFinished = mAttentionPriorCounterCloseToEnd >= 20;
```
Comment (tijyojwad) [inline cpp/include/tensorrt_llm/batch_manager/llmRequest.h]: nit: can be `return mAttentionPriorCounterCloseToEnd >= 20`

File: cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp [hunk 6]
```diff
@@ -957,20 +996,30 @@ void RuntimeBuffers::setAttentionPriorIdx(
...
-
     // create a cpu buffer for scores to find max score in
     SizeType32 searchLength = 10;
     auto const& manager = runtime.getBufferManager();
     auto const& stream = runtime.getStream();
     auto scoresHost = manager.cpu(ITensor::makeShape({searchLength}), scores->getDataType());
 
+    SizeType32 qOffset = 0;
+    // set the focus for context requests based on last scores slice
+    for (SizeType32 i = 0; i < (SizeType32)contextRequests.size(); ++i) {
```
Comment (tijyojwad) [inline cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp]: nit: i mainly use static_cast<type> when possible

File: tensorrt_llm/models/t5tts/model.py [hunk 7]
```diff
@@ -549,9 +554,21 @@ def forward(self,
...
+            # TODO: this is ones_like operation, is there more efficient way to do it?
+            cross_attention_mask = add(
+                mul(
+                    cross_attention_mask.cast(dtype=trt.float32),
+                    expand_dims_like(float(0), cross_attention_mask)
+                ),
+                expand_dims_like(float(1), cross_attention_mask)
```
Comment (tijyojwad) [inline tensorrt_llm/models/t5tts/model.py:567]: why do we need to to do mask * 0 + 1? why can't we just to `expand_dims_like(float(1), cross_attention_mask)` and use that as mask?

File: cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h (hunk 1 above)
Comment (tijyojwad) [reply cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h:2100]: not sure why this helps with attention mask?

Comment (vklimkov-nvidia) [reply cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h:2100]: for each element of qk tensor, if the attention_mask_ptr is set to 0, the -1e9 is added to the qk. when softmax is calculated for qk, those timestamps do not contribute to the context vector. This is inline with the rest of kernel template, where masking is done via negative bias, for example this:

```
relative_attention_bias += (FLT_MAX * (float(attention_mask_ptr[tlength]) - 1.0f));
``` 

this problem was that it was not happening for every timestamp. in the added line we actually do the masking operation for `local_time_now `

File: cpp/include/tensorrt_llm/batch_manager/llmRequest.h (hunk 4 above)
Comment (vklimkov-nvidia) [reply cpp/include/tensorrt_llm/batch_manager/llmRequest.h]: done

File: cpp/include/tensorrt_llm/batch_manager/llmRequest.h (hunk 5 above)
Comment (vklimkov-nvidia) [reply cpp/include/tensorrt_llm/batch_manager/llmRequest.h]: done

File: cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp (hunk 6 above)
Comment (vklimkov-nvidia) [reply cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp]: done

File: tensorrt_llm/models/t5tts/model.py (hunk 7 above)
Comment (vklimkov-nvidia) [reply tensorrt_llm/models/t5tts/model.py:567]: trying

Comment (vklimkov-nvidia) [reply tensorrt_llm/models/t5tts/model.py:567]: tried, but unfortunately didnt work. my guess would be is that tracing (making network graph out of python code) doesn't work as expected. in current implementation, resulting mask still "depends" on the cross_attention_mask even though its `cross_attention_mask * 0 + 1`. but in proposed simplified approach its an aux input and somehow it doesnt trace correctly.
